archunit/src/main/java/com/tngtech/archunit/library/cycle_detection/rules/CycleArchCondition.java

```diff
@@ -0,0 +1,382 @@
+                edges.addLast(edges.pollFirst());
+            }
+            Map<String, ComponentDependency<COMPONENT>> descriptionToEdge = new LinkedHashMap<>();
+            for (ComponentDependency<COMPONENT> edge : edges) {
+                descriptionToEdge.put(getDescriptionOfComponent.apply(edge.getOrigin()), edge);
+            }
+            return descriptionToEdge;
+        }
+
+        // A cycle always has edges, so we know that there is always at least one edge and by that a minimum element
+        // with respect to comparing the description lexicographically
+        @SuppressWarnings("OptionalGetWithoutIsPresent")
+        private ComponentDependency<COMPONENT> findStartEdge(Cycle<ComponentDependency<COMPONENT>> cycle) {
+            return cycle.getEdges().stream().min(comparing(input -> getDescriptionOfComponent.apply(input.getOrigin()))).get();
+        }
+
+        private String createDescription(Collection<String> edgeDescriptions, int indent) {
+            List<String> descriptions = new ArrayList<>(edgeDescriptions);
+            descriptions.add(descriptions.get(0));
+            return Joiner.on(" -> " + lineSeparator() + Strings.repeat(" ", indent)).join(descriptions);
+        }
+
+        private String createDetails(Map<String, ComponentDependency<COMPONENT>> descriptionsToEdges) {
+            List<String> details = new ArrayList<>();
+            int componentIndex = 0;
+            for (Map.Entry<String, ComponentDependency<COMPONENT>> edgeWithDescription : descriptionsToEdges.entrySet()) {
+                ++componentIndex;
+                details.add(String.format("  %d. Dependencies of %s", componentIndex, edgeWithDescription.getKey()));
+                details.addAll(dependenciesDescription(edgeWithDescription.getValue()));
+            }
+            return Joiner.on(lineSeparator()).join(details);
+        }
+
+        private List<String> dependenciesDescription(ComponentDependency<COMPONENT> edge) {
+            int maxDependencies = cycleConfiguration.getMaxNumberOfDependenciesToShowPerEdge();
+            Collection<Dependency> allDependencies = edge.toClassDependencies();
+            boolean tooManyDependenciesToDisplay = allDependencies.size() > maxDependencies;
+
+            Collection<Dependency> dependenciesToDisplay = tooManyDependenciesToDisplay
+                    ? allDependencies.stream().limit(maxDependencies).collect(toList())
+                    : allDependencies;
+
+            List<String> result = dependenciesToDisplay.stream()
+                    .map(dependency -> DEPENDENCY_DETAILS_INDENT + "- " + dependency.getDescription())
+                    .collect(toCollection(ArrayList::new));
+            if (tooManyDependenciesToDisplay) {
+                result.add(DEPENDENCY_DETAILS_INDENT + String.format("(%d further dependencies have been omitted...)",
+                        allDependencies.size() - dependenciesToDisplay.size()));
+            }
```

Comment:
Do you think that `Stream.limit` is expensive?
I'd drop the extra complexity to optimize the calculation of `dependenciesToDisplay` and just do:
```suggestion
            List<String> result = allDependencies.stream()
                    .limit(maxDependencies)
                    .map(dependency -> DEPENDENCY_DETAILS_INDENT + "- " + dependency.getDescription())
                    .collect(toCollection(ArrayList::new));
            if (tooManyDependenciesToDisplay) {
                result.add(DEPENDENCY_DETAILS_INDENT + String.format("(%d further dependencies have been omitted...)",
                        allDependencies.size() - maxDependencies));
            }
```

Reply:
Yeah, sounds reasonable, I didn't look over the code again because I just moved code from one place to the other in this refactoring. But sure, let's take the opportunity to simplify this a little :slightly_smiling_face: